File: 110.balanced-binary-tree.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=110 lang=cpp
 *
 * [110] Balanced Binary Tree
 */
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int calcHeight(TreeNode *node)
    {
        if (!node)
        {
            return 0;
        }

        int l = calcHeight(node->left);
        if (l == -1)
        {
            return -1;
        }

        int r = calcHeight(node->right);
        if (r == -1)
        {
            return -1;
        }

        if (abs(l - r) > 1)
        {
            return -1;
        }

        return max(l, r) + 1;
    }

    bool isBalanced(TreeNode *root)
    {
        return calcHeight(root) != -1;
    }
};
// @lc code=end

File: Java/Grind75/Week1/110.balanced-binary-tree.java
package Grind75.Week1;

/*
 * @lc app=leetcode id=110 lang=java
 *
 * [110] Balanced Binary Tree
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int treeHeight(TreeNode root) {
        if (root == null)
            return 0;
        int left = treeHeight(root.left), right = treeHeight(root.right);
        if (left != -1 && right != -1 && Math.abs(right - left) < 2)
            return Math.max(left, right) + 1;
        return -1;
    }

    public boolean isBalanced(TreeNode root) {
        if (treeHeight(root) == -1)
            return false;
        return true;
    }
}
// @lc code=end
